compiler-rt/lib/scudo/standalone/combined.h

```diff
@@ -1594,12 +1600,33 @@ class Allocator {
     RB->StackDepotSize = StackDepotSize;
     RB->RawStackDepotMap = DepotMap;
 
-    atomic_store(&RingBufferAddress, reinterpret_cast<uptr>(RB),
-                 memory_order_release);
+    swapOutRingBuffer(RB);
     static_assert(sizeof(AllocationRingBuffer) %
                           alignof(typename AllocationRingBuffer::Entry) ==
                       0,
                   "invalid alignment");
+    return true;
+  }
+
+  void swapOutRingBuffer(AllocationRingBuffer *NewRB) {
+    // To allow resizeRingBuffer to be called in a multi-threaded context by apps,
+    // we do not actually unmap, but only madvise(DONTNEED) the pages. That way,
+    // straggler threads will not crash.
```

Comment:
We don't leak the memory, we DONTNEED it so the kernel releases the actual memory.   We only leak a few pages in the case where other threads touch the pages after we unmapped it. We do leak the mapping, that is correct.

We would need to lock protect all the lock-free operations in order to disable and re-enable and make sure no thread currently has operations in progress.

In the case where we know we are single-threaded, we can unmap the buffers afterwards. This is to support the case where apps want to change the buffer (I forwarded you an email on why they might want that).

Reply:
In addition, keeping the old buffer doesn't prevent the potential crash. Any access to the old buffer can cause problem. Like if some thread tries to read the data from the old buffer which has been cleaned. Not to say the DONTNEED may happen at the same time some thread is still accessing it, which seems to be some undefined behavior. 

And yes, the buffer swap is complicated under current design (Thanks for the email BTW). One easy way in my mind is, we only reset the available size of ring buffer and the ring buffer will be initialized with some value is big enough for all of their cases. As a result, in most cases, they still use small memory for the ring buffer. 

I suppose we have the assumption that once we reset the available size, all the writes happened or happening will be viewed as staled (i.e., the buffer contains zero element). This will avoid the need of synchronizing the writing position and the available buffer size. 

I'm not sure if we need to do some changes here and there in Android. What do you think?

Reply:
> And yes, the buffer swap is complicated under current design (Thanks for the email BTW). One easy way in my mind is, we only reset the available size of ring buffer and the ring buffer will be initialized with some value is big enough for all of their cases. As a result, in most cases, they still use small memory for the ring buffer.

That is not possible with the current design of the StackDepot as a hash table.

Reply:
I'll take some time to review StackDepot and share the thought here later

Reply:
Sorry I was having some tasks on hands last two weeks. Will circle back to the issue this two days.

Reply:
Now the allocation ring buffer is allocated lazily. I'm thinking that we may not need to support actual resizing which will be complicated according to the current use cases (for example, the address of StackDepot is escaped. Which means we don't have an easy way to invalidate it)

To achieve the goal which wants to have a different ring buffer size for the different processes, instead of resizing the buffer, we can simply delay the ring buffer initialization. So it will be like,

1. Set the ring buffer size before running actual work load
2. Call malloc_set_track_allocation_stacks(true) to initialize the ring buffer

Then we can have the different ring buffer size across processes and it requires the least effort.

Reply:
I don't see how we would do this for apps. That would require all apps to call `malloc_set_track_allocation_stacks` themselves, otherwise they don't get any. Currently, the platform does this for them, but we wouldn't know whether we should do that, or the app wants to change the ring buffer size and thus call it itself.

Reply:
I would think something like for apps don't need to change the buffer size, they adopt the default behavior. Otherwise, they need to set the buffer size and call `malloc_set_track_allocation_stacks` themselves.

Before the app starts, I think there some places we can call `malloc_set_track_allocation_stacks` and maybe we can read the `AndroidManifest.xml` to determine if we want to change the default ring buffer size (Sorry I'm not familiar with app, not sure if AndroidManifest.xml can be used for this purpose)

We need some changes for native processes as well (like a different place to call `malloc_set_track_allocation_stacks`)

Overall, I would guess it's still doable but may require some additional works in Android. I can ask around to see other's opinions.

Reply:
An app developer explicitly asked for this to be a mallopt, not a manifest attribute.

Reply:
This DONTNEED usage looks safe to me. A trailing access with either repopulate the page or substitute a zero page, according to the spec. Not completely sure, but the latter might be the only actual possibility in Linux. In any case, the worst that may happen is we waste (leak) a page or two.

Reply:
The case I'm concerning is like, one thread is writing to the old buffer and another thread is calling DONTNEED on the old buffer. Both are writing ("either repopulate the page or substitute a zero page" is still a writing) to the same memory and they are not atomic operations. This is why I think it's an undefined behavior. 

About the wasting (leaking) pages, right, this is another concern I have. Even it's only for debugging, intentionally leaking memory in a memory allocator seems weird to me. In the perspective of maintenance, it's confusing  people who are not familiar with this.

I know it's unlikely to cause problem by this but I would prefer considering any approaches that are safer and reasonable.

Reply:
I agree that the spec says that the data is either repopulated or reads back as zero. However, if multiple threads are accessing this at the same time, I would presume that the repopulate or read back zero is on a page by page basis. Is there something in the kernel that makes the operation atomic for the entire mapped data? The kernel might make this atomic by preventing all interrupts while executing the madvise.

Honestly, if this is only really used for debugging purposes, it might be better to simply let it leak away and skip the madvise completely. Maybe even log that you are leaking away the old data. Even the data going to zero requires that anyone ever touching this code has to remember that is a possibility, and might result in undetected bugs in the future.

The best way would have been to copy the data for callers rather than having pointers to it. But since this is really a debugging item, a simpler interface with some known issues is probably fine.